File: variants/dotnet-csharp/Coordinator.cs
// =============================================================================
// Coordinator.cs — Coordenador (plano de controle) da variante .NET/C#.
//
// Espelha server.py + node_registry.py + metadata_service.py +
// replication_watcher.py. Cérebro do sistema: metadados (JSON), registro de nós
// com máquina de estados de vivacidade (ALIVE/SUSPECT/DEAD), placement
// determinístico, supervisor de re-replicação e garbage collection.
// O coordenador NUNCA toca nos bytes dos arquivos do usuário.
// =============================================================================

using System.Text.Json;
using System.Text.Json.Nodes;

namespace Dfs;

public class Coordinator
{
    private const string Alive = "ALIVE";
    private const string Suspect = "SUSPECT";
    private const string Dead = "DEAD";

    private class NodeState { public string Host = ""; public int Port; public double LastHb; public List<string> Chunks = new(); }
    public class Replica { public string node_id { get; set; } = ""; public string host { get; set; } = ""; public int port { get; set; } }
    public class ChunkEntry { public int index { get; set; } public string chunk_id { get; set; } = ""; public List<Replica> replicas { get; set; } = new(); }
    public class FileEntry { public int num_chunks { get; set; } public long chunk_size { get; set; } public double created_at { get; set; } public List<ChunkEntry> chunks { get; set; } = new(); }
    public class MetadataDoc { public Dictionary<string, FileEntry> files { get; set; } = new(); }

    private readonly object _lock = new();
    private readonly Dictionary<string, NodeState> _nodes = new();
    private readonly Dictionary<string, Dictionary<string, int>> _suspectedOrphans = new();
    private readonly Dictionary<string, List<string>> _pendingDeletes = new();
    private readonly Dictionary<string, string> _prevState = new();
    private int _fileCounter;

    private List<string> _membership;
    private readonly Dictionary<string, Config.NodeInfo> _nodeConfig = new();
    private MetadataDoc _metadata;

    private static readonly JsonSerializerOptions JsonOpts = new() { WriteIndented = true };

    public Coordinator()
    {
        _membership = Config.NodeOrder();
        foreach (var n in Config.BuildNodes()) _nodeConfig[n.NodeId] = n;
        Directory.CreateDirectory(Config.MetadataDir);
        _metadata = LoadMetadata();
    }

    public void Run()
    {
        StartWatcher();
        Log($"coordenador no ar em {Config.CoordinatorHost}:{Config.CoordinatorPort}");
        Log($"membership canônica: {string.Join(", ", _membership)} (RF={Config.ReplicationFactor}, quórum={Config.WriteQuorum})");
        Protocol.Serve(Config.CoordinatorHost, Config.CoordinatorPort, Handle);
    }

    // ---- Roteamento das chamadas do plano de controle -----------------------
    private JsonObject Handle(JsonObject req)
    {
        return Protocol.Str(req, "op") switch
        {
            "REGISTER" => HandleRegister(req),
            "HEARTBEAT" => HandleHeartbeat(req),
            "REQUEST_UPLOAD" => HandleRequestUpload(req),
            "CONFIRM_UPLOAD" => HandleConfirmUpload(req),
            "REQUEST_DOWNLOAD" => HandleRequestDownload(req),
            "DELETE_FILE" => HandleDeleteFile(req),
            "LIST_FILES" => HandleListFiles(),
            "UPDATE_REPLICAS" => HandleUpdateReplicas(req),
            "STATUS" => HandleStatus(),
            _ => new JsonObject { ["ok"] = false, ["error"] = $"op desconhecida: {Protocol.Str(req, "op")}" }
        };
    }

    private JsonObject HandleRegister(JsonObject req)
    {
        var nid = Protocol.Str(req, "node_id");
        lock (_lock)
        {
            _nodes[nid] = new NodeState { Host = Protocol.Str(req, "host"), Port = Protocol.Int(req, "port"), LastHb = Now() };
            if (!_membership.Contains(nid)) _membership.Add(nid);
            if (!_nodeConfig.ContainsKey(nid))
                _nodeConfig[nid] = new Config.NodeInfo(nid, Protocol.Str(req, "host"), Protocol.Int(req, "port"), "");
            _membership = Placement.SortNodes(_membership);
        }
        Log($"nó {nid} registrado (membership={_membership.Count})");
        return new JsonObject { ["ok"] = true, ["cluster_size"] = _membership.Count };
    }

    private JsonObject HandleHeartbeat(JsonObject req)
    {
        var nid = Protocol.Str(req, "node_id");
        var toDelete = new List<string>();
        lock (_lock)
        {
            if (!_nodes.TryGetValue(nid, out var node))
            {
                node = new NodeState { Host = _nodeConfig.GetValueOrDefault(nid)?.Host ?? "", Port = _nodeConfig.GetValueOrDefault(nid)?.Port ?? 0 };
                _nodes[nid] = node;
            }
            node.LastHb = Now();
            node.Chunks = (req["chunks"] as JsonArray)?.Select(x => x!.GetValue<string>()).ToList() ?? new List<string>();

            // Deleções pendentes (DELETE que falhou enquanto o nó estava morto).
            if (_pendingDeletes.TryGetValue(nid, out var pend)) { toDelete.AddRange(pend); _pendingDeletes.Remove(nid); }

            // Órfãos por block report, confirmados em 2 ciclos consecutivos.
            var expected = ExpectedChunksFor(nid);
            var seen = new HashSet<string>(node.Chunks);
            if (!_suspectedOrphans.TryGetValue(nid, out var suspects)) { suspects = new(); _suspectedOrphans[nid] = suspects; }
            var currentOrphans = new HashSet<string>(seen);
            currentOrphans.ExceptWith(expected);
            foreach (var cid in suspects.Keys.ToList())
                if (!currentOrphans.Contains(cid)) suspects.Remove(cid);
            foreach (var cid in currentOrphans)
            {
                suspects[cid] = suspects.GetValueOrDefault(cid) + 1;
                if (suspects[cid] >= 2) toDelete.Add(cid);
            }
        }
        return new JsonObject { ["ok"] = true, ["delete"] = ToJsonArray(toDelete.Distinct()) };
    }

    private JsonObject HandleRequestUpload(JsonObject req)
    {
        long size = Protocol.Long(req, "size");
        var uploadId = "up_" + Guid.NewGuid().ToString("N").Substring(0, 12);
        long chunkSize = ChooseChunkSize(size, _membership.Count);
        int numChunks = size <= 0 ? 1 : (int)Math.Ceiling((double)size / chunkSize);
        if (numChunks == 0) numChunks = 1;

        lock (_lock) { _fileCounter++; }

        var chunksArr = new JsonArray();
        for (int i = 0; i < numChunks; i++)
        {
            var replicaIds = Placement.ReplicasForChunk(i, _membership, Config.ReplicationFactor, _membership.Count);
            int live = replicaIds.Count(rid => StateOf(rid) != Dead);
            if (live < Config.WriteQuorum)
                return new JsonObject { ["ok"] = false, ["error"] = $"réplicas vivas insuficientes p/ quórum no chunk {i}" };

            var replicasJson = new JsonArray();
            foreach (var rid in replicaIds) replicasJson.Add(AddrJson(rid));
            chunksArr.Add(new JsonObject
            {
                ["index"] = i,
                ["chunk_id"] = $"{uploadId}_chunk_{i}",
                ["replicas"] = replicasJson,
                ["live_count"] = live
            });
        }

        Log($"upload {uploadId} p/ {Protocol.Str(req, "path")}: {numChunks} chunk(s) de {chunkSize} B");
        return new JsonObject
        {
            ["ok"] = true, ["upload_id"] = uploadId, ["chunk_size"] = chunkSize,
            ["num_chunks"] = numChunks, ["chunks"] = chunksArr
        };
    }

    private JsonObject HandleConfirmUpload(JsonObject req)
    {
        var path = Protocol.Str(req, "path");
        var entry = new FileEntry
        {
            chunk_size = Protocol.Long(req, "chunk_size"),
            created_at = Now(),
            chunks = new List<ChunkEntry>()
        };
        foreach (var c in (JsonArray)req["chunks"]!)
        {
            var co = (JsonObject)c!;
            var ce = new ChunkEntry { index = Protocol.Int(co, "index"), chunk_id = Protocol.Str(co, "chunk_id") };
            foreach (var r in (JsonArray)co["replicas"]!)
            {
                var ro = (JsonObject)r!;
                ce.replicas.Add(new Replica { node_id = Protocol.Str(ro, "node_id"), host = Protocol.Str(ro, "host"), port = Protocol.Int(ro, "port") });
            }
            entry.chunks.Add(ce);
        }
        entry.num_chunks = entry.chunks.Count;
        lock (_lock) { _metadata.files[path] = entry; PersistMetadata(); }
        Log($"arquivo {path} confirmado nos metadados");
        return new JsonObject { ["ok"] = true };
    }

    private JsonObject HandleRequestDownload(JsonObject req)
    {
        var path = Protocol.Str(req, "path");
        FileEntry? entry;
        lock (_lock) { _metadata.files.TryGetValue(path, out entry); }
        if (entry == null) return new JsonObject { ["ok"] = false, ["error"] = "arquivo não encontrado" };

        var chunksArr = new JsonArray();
        foreach (var c in entry.chunks)
        {
            var live = c.replicas.Where(r => StateOf(r.node_id) != Dead).ToList();
            var chosen = live.Count > 0 ? live : c.replicas;
            var replicasJson = new JsonArray();
            foreach (var r in chosen) replicasJson.Add(ReplicaJson(r));
            chunksArr.Add(new JsonObject { ["index"] = c.index, ["chunk_id"] = c.chunk_id, ["replicas"] = replicasJson });
        }
        return new JsonObject { ["ok"] = true, ["num_chunks"] = entry.num_chunks, ["chunks"] = chunksArr };
    }

    private JsonObject HandleDeleteFile(JsonObject req)
    {
        var path = Protocol.Str(req, "path");
        FileEntry? entry;
        lock (_lock) { _metadata.files.TryGetValue(path, out entry); }
        if (entry == null) return new JsonObject { ["ok"] = false, ["error"] = "arquivo não encontrado" };

        foreach (var c in entry.chunks)
        {
            foreach (var r in c.replicas)
            {
                if (StateOf(r.node_id) == Dead)
                    lock (_lock) { AddPending(r.node_id, c.chunk_id); }
                else
                {
                    try { Protocol.Request(r.host, r.port, new JsonObject { ["op"] = "DELETE", ["chunk_id"] = c.chunk_id }); }
                    catch { lock (_lock) { AddPending(r.node_id, c.chunk_id); } }
                }
            }
        }
        lock (_lock) { _metadata.files.Remove(path); PersistMetadata(); }
        Log($"arquivo {path} removido");
        return new JsonObject { ["ok"] = true };
    }

    private JsonObject HandleListFiles()
    {
        var arr = new JsonArray();
        lock (_lock)
        {
            foreach (var (path, e) in _metadata.files)
            {
                var nodes = e.chunks.SelectMany(c => c.replicas.Select(r => r.node_id)).Distinct().OrderBy(x => x).ToList();
                arr.Add(new JsonObject { ["path"] = path, ["num_chunks"] = e.num_chunks, ["nodes"] = ToJsonArray(nodes) });
            }
        }
        return new JsonObject { ["ok"] = true, ["files"] = arr };
    }

    private JsonObject HandleUpdateReplicas(JsonObject req)
    {
        lock (_lock)
        {
            if (_metadata.files.TryGetValue(Protocol.Str(req, "path"), out var entry))
            {
                var chunk = entry.chunks.FirstOrDefault(c => c.chunk_id == Protocol.Str(req, "chunk_id"));
                if (chunk != null && req["replicas"] is JsonArray ra)
                {
                    chunk.replicas = ra.Select(r => new Replica
                    {
                        node_id = Protocol.Str((JsonObject)r!, "node_id"),
                        host = Protocol.Str((JsonObject)r!, "host"),
                        port = Protocol.Int((JsonObject)r!, "port")
                    }).ToList();
                    PersistMetadata();
                }
            }
        }
        return new JsonObject { ["ok"] = true };
    }

    private JsonObject HandleStatus()
    {
        var arr = new JsonArray();
        foreach (var nid in _membership) arr.Add(new JsonObject { ["node_id"] = nid, ["state"] = StateOf(nid) });
        return new JsonObject { ["ok"] = true, ["nodes"] = arr, ["files"] = _metadata.files.Count };
    }

    // ---- Vivacidade (máquina de 3 estados, cálculo preguiçoso) --------------
    private string StateOf(string nodeId)
    {
        lock (_lock)
        {
            if (!_nodes.TryGetValue(nodeId, out var node) || node.LastHb == 0) return Dead;
            double silence = Now() - node.LastHb;
            if (silence < Config.HeartbeatSuspect) return Alive;
            if (silence < Config.HeartbeatDead) return Suspect;
            return Dead;
        }
    }

    private JsonObject AddrJson(string nodeId)
    {
        string host; int port;
        lock (_lock)
        {
            if (_nodes.TryGetValue(nodeId, out var n)) { host = n.Host; port = n.Port; }
            else if (_nodeConfig.TryGetValue(nodeId, out var c)) { host = c.Host; port = c.Port; }
            else { host = ""; port = 0; }
        }
        return new JsonObject { ["node_id"] = nodeId, ["host"] = host, ["port"] = port };
    }

    private static JsonObject ReplicaJson(Replica r) =>
        new() { ["node_id"] = r.node_id, ["host"] = r.host, ["port"] = r.port };

    private HashSet<string> ExpectedChunksFor(string nodeId)
    {
        var set = new HashSet<string>();
        foreach (var e in _metadata.files.Values)
            foreach (var c in e.chunks)
                if (c.replicas.Any(r => r.node_id == nodeId)) set.Add(c.chunk_id);
        return set;
    }

    private void AddPending(string nodeId, string chunkId)
    {
        if (!_pendingDeletes.TryGetValue(nodeId, out var list)) { list = new(); _pendingDeletes[nodeId] = list; }
        list.Add(chunkId);
    }

    // ---- Supervisor de re-replicação ----------------------------------------
    private void StartWatcher()
    {
        var t = new Thread(() =>
        {
            while (true)
            {
                Thread.Sleep(Config.WatcherInterval * 1000);
                try { DetectAndHeal(); }
                catch (Exception e) { Log($"watcher erro: {e.Message}"); }
            }
        }) { IsBackground = true };
        t.Start();
    }

    private void DetectAndHeal()
    {
        var transitions = new List<string>();
        lock (_lock)
        {
            foreach (var nid in _membership)
            {
                var cur = StateOf(nid);
                if (cur == Dead && _prevState.GetValueOrDefault(nid) != Dead) transitions.Add(nid);
                _prevState[nid] = cur;
            }
        }
        if (transitions.Count == 0) return;
        foreach (var nid in transitions) Log($"detectada MORTE de {nid}: iniciando re-replicação");

        var work = new List<(string path, ChunkEntry chunk, string source, string target)>();
        lock (_lock)
        {
            foreach (var (path, e) in _metadata.files)
            {
                foreach (var c in e.chunks)
                {
                    var replicaIds = c.replicas.Select(r => r.node_id).ToList();
                    var dead = replicaIds.Where(rid => StateOf(rid) == Dead).ToList();
                    if (dead.Count == 0) continue;
                    var live = replicaIds.Where(rid => StateOf(rid) != Dead).ToList();
                    if (live.Count == 0) continue;
                    int need = Config.ReplicationFactor - live.Count;
                    if (need <= 0) continue;

                    var candidates = _membership.Where(m => StateOf(m) != Dead && !replicaIds.Contains(m)).Take(need);
                    foreach (var target in candidates)
                        work.Add((path, c, live[0], target));
                }
            }
        }

        foreach (var w in work) ReplicateChunk(w.chunk, w.source, w.target);
    }

    private void ReplicateChunk(ChunkEntry chunk, string source, string target)
    {
        try
        {
            var src = AddrJson(source);
            var tgt = AddrJson(target);
            var resp = Protocol.Request(Protocol.Str(src, "host"), Protocol.Int(src, "port"),
                new JsonObject { ["op"] = "REPLICATE", ["chunk_id"] = chunk.chunk_id, ["target"] = tgt });
            if (!Protocol.Bool(resp, "ok")) return;

            lock (_lock)
            {
                chunk.replicas = chunk.replicas.Where(r => StateOf(r.node_id) != Dead).ToList();
                chunk.replicas.Add(new Replica { node_id = target, host = Protocol.Str(tgt, "host"), port = Protocol.Int(tgt, "port") });
                chunk.replicas = chunk.replicas.GroupBy(r => r.node_id).Select(g => g.First()).ToList();
                PersistMetadata();
            }
            Log($"chunk {chunk.chunk_id} re-replicado {source} -> {target}");
        }
        catch (Exception e) { Log($"falha ao re-replicar {chunk.chunk_id}: {e.Message}"); }
    }

    // ---- Tamanho de chunk adaptável (porta de chunking.py) ------------------
    private static long ChooseChunkSize(long fileSize, int clusterSize)
    {
        if (fileSize <= 0) return Config.MinChunkSize;
        long candidate = fileSize / (clusterSize * Config.ChunkTargetMultiplier);
        if (fileSize >= clusterSize * Config.MinChunkSize) candidate = Math.Min(candidate, fileSize / clusterSize);
        return Math.Max(Config.MinChunkSize, Math.Min(candidate, Config.MaxChunkSize));
    }

    // ---- Persistência de metadados ------------------------------------------
    private MetadataDoc LoadMetadata()
    {
        try
        {
            if (File.Exists(Config.MetadataFile))
                return JsonSerializer.Deserialize<MetadataDoc>(File.ReadAllText(Config.MetadataFile)) ?? new MetadataDoc();
        }
        catch { /* índice corrompido: começa vazio */ }
        return new MetadataDoc();
    }

    private void PersistMetadata() =>
        File.WriteAllText(Config.MetadataFile, JsonSerializer.Serialize(_metadata, JsonOpts));

    private static double Now() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

    private static JsonArray ToJsonArray(IEnumerable<string> items)
    {
        var arr = new JsonArray();
        foreach (var s in items) arr.Add(s);
        return arr;
    }

    private void Log(string msg) => Console.WriteLine($"[coordenador] {msg}");
}

File: variants/ruby/coordinator.rb
# frozen_string_literal: true

# =============================================================================
# coordinator.rb — Coordenador (plano de controle) da variante Ruby.
#
# Espelha server.py + node_registry.py + metadata_service.py +
# replication_watcher.py do original. Cérebro do sistema:
#   - METADADOS (arquivos -> chunks -> réplicas), persistidos em JSON.
#   - REGISTRO DE NÓS com máquina de vivacidade (ALIVE/SUSPECT/DEAD) preguiçosa.
#   - PLACEMENT determinístico + escolha de INGRESS (round-robin entre nós vivos)
#     e de EGRESS (por localidade: o nó vivo com mais chunks do arquivo).
#   - SUPERVISOR DE RE-REPLICAÇÃO e GARBAGE COLLECTION por block report.
#   - TELEMETRIA: agrega métricas de upload/download reportadas pelos nós
#     ingress/egress (min/máx/média/contagem), no espírito do telemetry_hub.
#
# O coordenador NUNCA toca nos bytes dos arquivos do usuário — só metadados.
# =============================================================================

require 'json'
require 'set'
require 'securerandom'
require_relative 'config'
require_relative 'lib/protocol'
require_relative 'lib/placement'

module DFS
  class Coordinator
    include DFS::Config

    ALIVE = 'ALIVE'
    SUSPECT = 'SUSPECT'
    DEAD = 'DEAD'

    def initialize
      @mutex = Mutex.new
      @nodes = {}            # node_id => { host, port, last_hb, chunks:[] }
      @suspected_orphans = Hash.new { |h, k| h[k] = {} } # node_id => {chunk_id=>count}
      @pending_deletes = Hash.new { |h, k| h[k] = [] }   # node_id => [chunk_id]
      @prev_state = {}       # node_id => estado anterior (p/ detectar transições)
      @upload_counter = 0    # contador monotônico p/ round-robin de ingress

      # Telemetria: agregados por operação + contadores de eventos.
      @metrics = Hash.new { |h, k| h[k] = { 'count' => 0, 'total' => 0.0, 'min' => nil, 'max' => nil, 'bytes' => 0 } }
      @rereplications = 0
      @gc_deletes = 0

      # Membership canônica: a lista OFICIAL dos N nós (base do placement).
      @membership = DFS::Config.node_order
      @node_config = {}
      DFS::Config.build_nodes.each { |n| @node_config[n['node_id']] = n }

      FileUtils.mkdir_p(METADATA_DIR)
      @metadata = load_metadata
    end

    def run
      start_watcher_thread
      log "coordenador no ar em #{COORDINATOR_HOST}:#{COORDINATOR_PORT}"
      log "membership canônica: #{@membership.join(', ')} (RF=#{REPLICATION_FACTOR}, quórum=#{WRITE_QUORUM})"
      DFS::Protocol.serve(COORDINATOR_HOST, COORDINATOR_PORT) { |req| handle(req) }
    end

    private

    # ---- Roteamento das chamadas do plano de controle -----------------------
    def handle(req)
      case req['op']
      when 'REGISTER'         then handle_register(req)
      when 'HEARTBEAT'        then handle_heartbeat(req)
      when 'REQUEST_UPLOAD'   then handle_request_upload(req)
      when 'CONFIRM_UPLOAD'   then handle_confirm_upload(req)
      when 'REQUEST_DOWNLOAD' then handle_request_download(req)
      when 'DELETE_FILE'      then handle_delete_file(req)
      when 'LIST_FILES'       then handle_list_files
      when 'UPDATE_REPLICAS'  then handle_update_replicas(req)
      when 'METRIC'           then handle_metric(req)
      when 'METRICS'          then handle_metrics
      when 'STATUS'           then handle_status
      else { 'ok' => false, 'error' => "op desconhecida: #{req['op']}" }
      end
    end

    # RegisterNode: promove o nó à membership e registra seu endereço.
    def handle_register(req)
      @mutex.synchronize do
        nid = req['node_id']
        @nodes[nid] = { 'host' => req['host'], 'port' => req['port'],
                        'last_hb' => Time.now.to_f, 'chunks' => [] }
        @membership << nid unless @membership.include?(nid)
        @node_config[nid] ||= { 'node_id' => nid, 'host' => req['host'], 'port' => req['port'] }
        @membership = DFS::Placement.sort_nodes(@membership)
      end
      log "nó #{req['node_id']} registrado (membership=#{@membership.size})"
      { 'ok' => true, 'cluster_size' => @membership.size, 'membership' => @membership }
    end

    # Heartbeat: renova a vivacidade, guarda o block report e devolve a lista de
    # órfãos a apagar (deleções pendentes + detecção por dois ciclos).
    def handle_heartbeat(req)
      nid = req['node_id']
      to_delete = []
      @mutex.synchronize do
        node = (@nodes[nid] ||= { 'host' => @node_config.dig(nid, 'host'),
                                  'port' => @node_config.dig(nid, 'port'), 'chunks' => [] })
        node['last_hb'] = Time.now.to_f
        node['chunks'] = Array(req['chunks'])

        pend = @pending_deletes.delete(nid) || []
        to_delete.concat(pend)

        expected = expected_chunks_for(nid)
        seen = node['chunks'].to_set
        suspects = @suspected_orphans[nid]
        current_orphans = seen - expected
        suspects.each_key { |cid| suspects.delete(cid) unless current_orphans.include?(cid) }
        current_orphans.each do |cid|
          suspects[cid] = (suspects[cid] || 0) + 1
          to_delete << cid if suspects[cid] >= 2
        end
        @gc_deletes += to_delete.size
      end
      { 'ok' => true, 'delete' => to_delete.uniq }
    end

    # RequestUpload: calcula chunk_size, placement de cada chunk (membership
    # canônica) e escolhe o INGRESS entre os nós VIVOS (round-robin por arquivo).
    # O cliente NÃO fatia nada — quem fatia e replica é o ingress.
    def handle_request_upload(req)
      size = req['size'].to_i
      upload_id = "up_#{SecureRandom.hex(6)}"
      chunk_size = choose_chunk_size(size, @membership.size)
      num_chunks = size <= 0 ? 1 : (size.to_f / chunk_size).ceil
      num_chunks = 1 if num_chunks.zero?

      chunks = (0...num_chunks).map do |i|
        replica_ids = DFS::Placement.replicas_for_chunk(
          i, @membership, REPLICATION_FACTOR, cluster_size: @membership.size
        )
        live = replica_ids.select { |rid| state_of(rid) != DEAD }
        { 'index' => i, 'chunk_id' => "#{upload_id}_chunk_#{i}",
          'replicas' => replica_ids.map { |rid| addr(rid) }, 'live_count' => live.size }
      end

      insufficient = chunks.find { |c| c['live_count'] < WRITE_QUORUM }
      return { 'ok' => false, 'error' => "réplicas vivas insuficientes p/ quórum no chunk #{insufficient['index']}" } if insufficient

      # Ingress: round-robin ENTRE os nós vivos (o placement usa a canônica).
      live_members = @membership.select { |m| state_of(m) != DEAD }
      return { 'ok' => false, 'error' => 'nenhum nó vivo para ingress' } if live_members.empty?

      idx = nil
      @mutex.synchronize { idx = @upload_counter; @upload_counter += 1 }
      ingress_id = DFS::Placement.sort_nodes(live_members)[idx % live_members.size]

      log "upload #{upload_id} p/ #{req['path']}: #{num_chunks} chunk(s) de #{chunk_size} B, ingress=#{ingress_id}"
      { 'ok' => true, 'upload_id' => upload_id, 'chunk_size' => chunk_size,
        'num_chunks' => num_chunks, 'ingress' => addr(ingress_id), 'chunks' => chunks }
    end

    # ConfirmUpload: registra o arquivo nos metadados com as réplicas efetivas.
    # É chamado pelo INGRESS (que sabe o que gravou), não pelo cliente fraco.
    def handle_confirm_upload(req)
      @mutex.synchronize do
        @metadata['files'][req['path']] = {
          'num_chunks' => req['chunks'].size,
          'chunk_size' => req['chunk_size'],
          'size' => req['size'],
          'created_at' => Time.now.to_f,
          'chunks' => req['chunks'].map do |c|
            { 'index' => c['index'], 'chunk_id' => c['chunk_id'], 'replicas' => c['replicas'] }
          end
        }
        persist_metadata
      end
      log "arquivo #{req['path']} confirmado nos metadados (ingress=#{req['ingress']})"
      { 'ok' => true }
    end

    # RequestDownload: devolve o mapa de chunks e escolhe o EGRESS por LOCALIDADE
    # (o nó vivo que já tem o maior número de chunks do arquivo).
    def handle_request_download(req)
      entry = nil
      @mutex.synchronize { entry = @metadata['files'][req['path']] }
      return { 'ok' => false, 'error' => 'arquivo não encontrado' } unless entry

      chunks = entry['chunks'].map do |c|
        { 'index' => c['index'], 'chunk_id' => c['chunk_id'], 'replicas' => c['replicas'] }
      end

      counts = Hash.new(0)
      entry['chunks'].each do |c|
        c['replicas'].each { |r| counts[r['node_id']] += 1 if state_of(r['node_id']) != DEAD }
      end
      return { 'ok' => false, 'error' => 'nenhuma réplica viva para servir o download' } if counts.empty?

      egress_id = counts.max_by { |_nid, cnt| cnt }.first
      log "download #{req['path']}: egress=#{egress_id} (#{counts[egress_id]} de #{entry['num_chunks']} chunks locais)"
      { 'ok' => true, 'num_chunks' => entry['num_chunks'], 'egress' => addr(egress_id), 'chunks' => chunks }
    end

    def handle_delete_file(req)
      entry = nil
      @mutex.synchronize { entry = @metadata['files'][req['path']] }
      return { 'ok' => false, 'error' => 'arquivo não encontrado' } unless entry

      entry['chunks'].each do |c|
        c['replicas'].each do |r|
          rid = r['node_id']
          if state_of(rid) == DEAD
            @mutex.synchronize { @pending_deletes[rid] << c['chunk_id'] }
          else
            begin
              DFS::Protocol.request(r['host'], r['port'], 'op' => 'DELETE', 'chunk_id' => c['chunk_id'])
            rescue StandardError
              @mutex.synchronize { @pending_deletes[rid] << c['chunk_id'] }
            end
          end
        end
      end
      @mutex.synchronize do
        @metadata['files'].delete(req['path'])
        persist_metadata
      end
      log "arquivo #{req['path']} removido"
      { 'ok' => true }
    end

    def handle_list_files
      files = @mutex.synchronize do
        @metadata['files'].map do |path, e|
          nodes = e['chunks'].flat_map { |c| c['replicas'].map { |r| r['node_id'] } }.uniq.sort
          { 'path' => path, 'num_chunks' => e['num_chunks'], 'size' => e['size'], 'nodes' => nodes }
        end
      end
      { 'ok' => true, 'files' => files }
    end

    def handle_update_replicas(req)
      @mutex.synchronize do
        entry = @metadata['files'][req['path']]
        if entry
          chunk = entry['chunks'].find { |c| c['chunk_id'] == req['chunk_id'] }
          chunk['replicas'] = req['replicas'] if chunk
          persist_metadata
        end
      end
      { 'ok' => true }
    end

    # METRIC: um nó ingress/egress reporta a duração e o volume de uma operação.
    def handle_metric(req)
      op = req['metric'] # 'upload' | 'download'
      @mutex.synchronize do
        m = @metrics[op]
        dur = req['duration'].to_f
        m['count'] += 1
        m['total'] += dur
        m['bytes'] += req['bytes'].to_i
        m['min'] = dur if m['min'].nil? || dur < m['min']
        m['max'] = dur if m['max'].nil? || dur > m['max']
      end
      { 'ok' => true }
    end

    def handle_metrics
      @mutex.synchronize do
        ops = {}
        @metrics.each do |op, m|
          avg = m['count'].zero? ? 0.0 : m['total'] / m['count']
          ops[op] = { 'count' => m['count'], 'avg_ms' => (avg * 1000).round(2),
                      'min_ms' => ((m['min'] || 0) * 1000).round(2),
                      'max_ms' => ((m['max'] || 0) * 1000).round(2),
                      'bytes' => m['bytes'] }
        end
        { 'ok' => true, 'ops' => ops, 'rereplications' => @rereplications,
          'gc_deletes' => @gc_deletes, 'files' => @metadata['files'].size }
      end
    end

    def handle_status
      st = @membership.map { |nid| { 'node_id' => nid, 'state' => state_of(nid) } }
      { 'ok' => true, 'nodes' => st, 'files' => @metadata['files'].size,
        'rereplications' => @rereplications, 'gc_deletes' => @gc_deletes }
    end

    # ---- Vivacidade (máquina de 3 estados, cálculo preguiçoso) --------------
    def state_of(node_id)
      node = @nodes[node_id]
      return DEAD if node.nil? || node['last_hb'].nil?

      silence = Time.now.to_f - node['last_hb']
      if silence < HEARTBEAT_SUSPECT then ALIVE
      elsif silence < HEARTBEAT_DEAD then SUSPECT
      else DEAD
      end
    end

    def addr(node_id)
      n = @nodes[node_id] || @node_config[node_id] || {}
      { 'node_id' => node_id, 'host' => n['host'], 'port' => n['port'] }
    end

    def expected_chunks_for(node_id)
      set = Set.new
      @metadata['files'].each_value do |e|
        e['chunks'].each do |c|
          set << c['chunk_id'] if c['replicas'].any? { |r| r['node_id'] == node_id }
        end
      end
      set
    end

    # ---- Supervisor de re-replicação ----------------------------------------
    def start_watcher_thread
      Thread.new do
        loop do
          sleep WATCHER_INTERVAL
          begin
            detect_and_heal
          rescue StandardError => e
            log "watcher erro: #{e.message}"
          end
        end
      end
    end

    def detect_and_heal
      transitions = []
      @mutex.synchronize do
        @membership.each do |nid|
          cur = state_of(nid)
          transitions << nid if cur == DEAD && @prev_state[nid] != DEAD
          @prev_state[nid] = cur
        end
      end
      return if transitions.empty?

      transitions.each { |nid| log "detectada MORTE de #{nid}: iniciando re-replicação" }

      work = []
      @mutex.synchronize do
        @metadata['files'].each do |path, e|
          e['chunks'].each do |c|
            replica_ids = c['replicas'].map { |r| r['node_id'] }
            dead = replica_ids.select { |rid| state_of(rid) == DEAD }
            next if dead.empty?

            live = replica_ids.select { |rid| state_of(rid) != DEAD }
            next if live.empty?

            need = REPLICATION_FACTOR - live.size
            next if need <= 0

            candidates = @membership.select { |m| state_of(m) != DEAD && !replica_ids.include?(m) }
            candidates.first(need).each do |target|
              work << { path: path, chunk: c, source: live.first, target: target }
            end
          end
        end
      end

      work.each { |w| replicate_chunk(w) }
    end

    def replicate_chunk(w)
      src = addr(w[:source])
      tgt = addr(w[:target])
      resp = DFS::Protocol.request(src['host'], src['port'],
                                   'op' => 'REPLICATE', 'chunk_id' => w[:chunk]['chunk_id'],
                                   'target' => tgt)
      return unless resp['ok']

      @mutex.synchronize do
        new_replicas = w[:chunk]['replicas'].reject { |r| state_of(r['node_id']) == DEAD }
        new_replicas << tgt
        w[:chunk]['replicas'] = new_replicas.uniq { |r| r['node_id'] }
        @rereplications += 1
        persist_metadata
      end
      log "chunk #{w[:chunk]['chunk_id']} re-replicado #{w[:source]} -> #{w[:target]}"
    rescue StandardError => e
      log "falha ao re-replicar #{w[:chunk]['chunk_id']}: #{e.message}"
    end

    # ---- Tamanho de chunk adaptável (porta de chunking.py) ------------------
    def choose_chunk_size(file_size, cluster_size)
      return MIN_CHUNK_SIZE if file_size <= 0

      candidate = file_size / (cluster_size * CHUNK_TARGET_MULTIPLIER)
      candidate = [candidate, file_size / cluster_size].min if file_size >= cluster_size * MIN_CHUNK_SIZE
      [MIN_CHUNK_SIZE, [candidate, MAX_CHUNK_SIZE].min].max
    end

    # ---- Persistência de metadados ------------------------------------------
    def load_metadata
      if File.exist?(METADATA_FILE)
        JSON.parse(File.read(METADATA_FILE))
      else
        { 'files' => {} }
      end
    rescue StandardError
      { 'files' => {} }
    end

    def persist_metadata
      File.write(METADATA_FILE, JSON.pretty_generate(@metadata))
    end

    def log(msg)
      puts "[coordenador] #{msg}"
    end
  end
end

if $PROGRAM_NAME == __FILE__
  $stdout.sync = true
  DFS::Coordinator.new.run
end
